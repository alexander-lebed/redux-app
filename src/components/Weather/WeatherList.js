// @flow
import React from 'react';
import $http from 'axios';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Col from 'react-bootstrap/lib/Col';
import Row from 'react-bootstrap/lib/Row';
import ButtonToolbar from 'react-bootstrap/lib/ButtonToolbar';
import DropdownButton from 'react-bootstrap/lib/DropdownButton';
import MenuItem from 'react-bootstrap/lib/MenuItem';
import type { State, Dispatch } from '../../redux/types';
import { addLocation, updateLocations, removeLocation } from '../../redux/actions/weather-actions';
import Location from './Location';


type Props = {
    locations: State,
    addLocation: (id: string) => void,
    updateLocations: (locations: State) => void,
    removeLocation: (id: string) => void
}

export class WeatherList extends React.Component<void, Props, void> {
  componentDidMount() {
    this.updateWeather();
  }

  updateWeather() {
    const { locations, updateLocations } = this.props;
    const getWeatherUrl = (location): string => `
            https://query.yahooapis.com/v1/public/yql?q=
            select * from weather.forecast
            where woeid in (
                select woeid from geo.places(1)
                where text="${location.get('city')}, ${location.get('country')}"
            )
            &format=json`;
    const promises = locations.map(l => $http.get(getWeatherUrl(l)));
    Promise.all(promises).then(([...response]) => {
      const data = locations.map((l, index) => {
        const results = response[index].data.query.results;
        return l.set('temp', results ? results.channel.item.condition.temp : 0);
      });
      updateLocations(data);
    });
  }

  render() {
    const { locations, addLocation, removeLocation } = this.props;
    const onAddLocation = (location) => {
      addLocation(location.get('id'));
      setTimeout(this.updateWeather.bind(this), 50);
    };
    const deletedLocations = locations.filter(l => l.get('isDeleted'));
    return (
      <div className="todo">
        <Row>
          <Col xs={3} />
          <Col xs={6}>
            <ul className="todo__list">
              {locations.filter(l => !l.get('isDeleted'))
                .map(l => (
                  <li key={l.get('id')} className="todo__item">
                    <Location location={l} onDelete={() => removeLocation(l.get('id'))} />
                  </li>
                ))
              }
            </ul>
            <ButtonToolbar style={{ marginTop: 10 }}>
              <DropdownButton
                id="add-location"
                title="Add Location"
                disabled={deletedLocations.size === 0}
                onSelect={l => onAddLocation(l)}
              >
                {deletedLocations.map(l => (
                  <MenuItem key={l.get('id')} eventKey={l}>
                    {`${l.get('city')}, ${l.get('country')}`}
                  </MenuItem>
                ))}
              </DropdownButton>
            </ButtonToolbar>
          </Col>
          <Col xs={3} />
        </Row>
      </div>
    );
  }
}

export default connect(
  state => ({ locations: state.locations }),
  (dispatch: Dispatch) =>
    bindActionCreators({ addLocation, updateLocations, removeLocation }, dispatch)
)(WeatherList);
