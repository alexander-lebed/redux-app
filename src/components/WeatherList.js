// @flow
import React from 'react';
import $http from 'axios';
import {connect} from 'react-redux';
import {Map} from 'immutable';
import type {State, Dispatch} from '../types';
import Col from 'react-bootstrap/lib/Col';
import Row from 'react-bootstrap/lib/Row';
import Glyphicon from 'react-bootstrap/lib/Glyphicon';
import ButtonToolbar from 'react-bootstrap/lib/ButtonToolbar';
import DropdownButton from 'react-bootstrap/lib/DropdownButton';
import MenuItem from 'react-bootstrap/lib/MenuItem';
import {addLocation, updateLocations, removeLocation} from '../actions/weather-actions';


type LocationProps = {
    location: Map<string, any>,
    onDelete: (id: string) => void
}

type WeatherListProps = {
    locations: State,
    addLocation: (id: string) => void,
    updateLocations: (locations: State) => void,
    removeLocation: (id: string) => void
}

const Location = (props: LocationProps) => {
    const {location, onDelete} = props;
    let temperature = '';
    if (location.get('temp')) {
        const degreesInCelsius = Math.round((location.get('temp') - 32) / 1.8);
        temperature = `${degreesInCelsius} Â°C`;
    } else {
        temperature = 'not available';
    }
    return (
        <Row>
            <Col xs={6}>{location.get('city')}</Col>
            <Col xs={3}>{temperature}</Col>
            <Col xs={3}>
                <Glyphicon
                    glyph="remove"
                    className="pull-right"
                    style={{marginRight: 10}}
                    title="Remove Location"
                    onClick={onDelete}
                />
            </Col>
        </Row>
    );
};

class WeatherList extends React.Component<void, WeatherListProps, void> {

    componentDidMount () {
        this.updateWeather();
    }

    updateWeather () {
        const {locations, updateLocations} = this.props;
        const getWeatherUrl = (location): string => {
            return `
            https://query.yahooapis.com/v1/public/yql?q=
            select * from weather.forecast
            where woeid in (
                select woeid from geo.places(1)
                where text="${location.get('city')}, ${location.get('country')}"
            )
            &format=json`
        };
        const promises = locations.map(l => $http.get(getWeatherUrl(l)));
        Promise.all(promises).then(([...response]) => {
            const data = locations.map((l, index) => {
                const results = response[index].data.query.results
                return l.set('temp', results ? results.channel.item.condition.temp : 0);
            })
            updateLocations(data);
        })
    }

    render () {
        const {locations, addLocation, removeLocation} = this.props;
        const onDelete = (id: string) => event => removeLocation(id);
        const onAddLocation = (location) => {
            addLocation(location.get('id'));
            setTimeout(this.updateWeather.bind(this), 0);
        }
        const deletedLocations = locations.filter(l => l.get('isDeleted'))
        return (
            <div className="todo">
                <Row>
                    <Col xs={3} />
                    <Col xs={6}>
                        <ul className="todo__list">
                            {locations.filter(l => !l.get('isDeleted'))
                                .map(l => (
                                    <li key={l.get('id')} className="todo__item">
                                        <Location location={l} onDelete={onDelete(l.get('id'))} />
                                    </li>
                                ))
                            }
                        </ul>
                        <ButtonToolbar style={{marginTop: 10}}>
                            <DropdownButton
                                id="add-location"
                                title="Add Location"
                                disabled={deletedLocations.size === 0}
                                onSelect={(evt) => onAddLocation(evt)}
                            >
                                {deletedLocations.map(l => (
                                    <MenuItem key={l.get('id')} eventKey={l}>
                                        {`${l.get('city')}, ${l.get('country')}`}
                                    </MenuItem>
                                ))}
                            </DropdownButton>
                        </ButtonToolbar>
                    </Col>
                    <Col xs={3} />
                </Row>
            </div>
        )
    }
}

export default connect(
    (state) => {
        return {
            locations: state.locations
        }
    },
    (dispatch: Dispatch) => {
        return {
            addLocation: id => dispatch(addLocation(id)),
            updateLocations: locations => dispatch(updateLocations(locations)),
            removeLocation: id => dispatch(removeLocation(id))
        }
    }
)(WeatherList)