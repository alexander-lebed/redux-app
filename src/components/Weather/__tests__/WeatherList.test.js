// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { List, Map } from 'immutable';
import { WeatherList } from '../WeatherList';


describe('<WeatherList/>', () => {
    test('should render WeatherList', () => {
        const locs = List([
            Map({ id: 1, city: 'Odessa', country: 'Ukraine', temp: 0, isDeleted: false }),
            Map({ id: 2, city: 'New York', country: 'United States', temp: 0, isDeleted: false })
        ]);
        const wrapper = renderer.create(
            <WeatherList
                locations={locs}
                addLocation={jest.fn()}
                updateLocations={jest.fn()}
                removeLocation={jest.fn()}
            />
        );
        const tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
