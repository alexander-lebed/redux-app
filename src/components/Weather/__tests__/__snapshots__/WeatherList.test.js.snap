// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherList/> should match snapshot 1`] = `
<div
  className="row"
  style={
    Object {
      "marginLeft": 0,
      "marginRight": 0,
    }
  }
>
  <div
    className="col-sm-6 col-sm-offset-3 col-xs-12 col-xs-offset-0"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-6 col-xs-12"
      >
        <div
          onKeyUp={[Function]}
        >
          <div
            className="Select is-clearable is-searchable Select--single"
            style={undefined}
          >
            <div
              className="Select-control"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={undefined}
            >
              <span
                className="Select-multi-value-wrapper"
                id="react-select-2--value"
              >
                <div
                  className="Select-placeholder"
                >
                  Search for a city
                </div>
                <div
                  className="Select-input"
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <input
                    aria-activedescendant="react-select-2--value"
                    aria-describedby={undefined}
                    aria-expanded="false"
                    aria-haspopup="false"
                    aria-label={undefined}
                    aria-labelledby={undefined}
                    aria-owns=""
                    className={undefined}
                    id={undefined}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    required={false}
                    role="combobox"
                    style={
                      Object {
                        "boxSizing": "content-box",
                        "width": "5px",
                      }
                    }
                    tabIndex={undefined}
                    value=""
                  />
                  <div
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "overflow": "scroll",
                        "position": "absolute",
                        "top": 0,
                        "visibility": "hidden",
                        "whiteSpace": "pre",
                      }
                    }
                  >
                    
                  </div>
                </div>
              </span>
              <span
                className="Select-arrow-zone"
                onMouseDown={[Function]}
              >
                <span
                  className="Select-arrow"
                  onMouseDown={[Function]}
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "paddingTop": 15,
        }
      }
    >
      <ul
        className="weather-list"
      >
        <li
          className="weather-item"
        >
          <div
            className="row"
            style={
              Object {
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <div
              className="col-xs-6"
            >
              <span
                className="glyphicon glyphicon-map-marker"
                style={
                  Object {
                    "color": "#00b386",
                    "marginRight": 5,
                  }
                }
              />
              Odessa (OD), Ukraine
            </div>
            <div
              className="col-xs-4"
            >
              not available
            </div>
            <div
              className="col-xs-2"
            >
              <span
                className="pull-right glyphicon glyphicon-trash"
                onClick={[Function]}
                style={
                  Object {
                    "color": "grey",
                    "marginRight": 10,
                  }
                }
                title="Remove Location"
              />
            </div>
          </div>
        </li>
        <li
          className="weather-item"
        >
          <div
            className="row"
            style={
              Object {
                "paddingBottom": 5,
                "paddingTop": 5,
              }
            }
          >
            <div
              className="col-xs-6"
            >
              <span
                className="glyphicon glyphicon-map-marker"
                style={
                  Object {
                    "color": "#00b386",
                    "marginRight": 5,
                  }
                }
              />
              New York (NY), United States
            </div>
            <div
              className="col-xs-4"
            >
              not available
            </div>
            <div
              className="col-xs-2"
            >
              <span
                className="pull-right glyphicon glyphicon-trash"
                onClick={[Function]}
                style={
                  Object {
                    "color": "grey",
                    "marginRight": 10,
                  }
                }
                title="Remove Location"
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
